REGRESSION (279833@main): webrtc/vp8-then-h264-gpu-process-crash.html is consistently crashing

kkinnunen-apple opened the pull request:
#### b7f69a1c3709ddda04cd84b14dfb1068a93b6865
<pre>
REGRESSION (279833@main): webrtc/vp8-then-h264-gpu-process-crash.html is consistently crashing
<a href="https://bugs.webkit.org/show_bug.cgi?id=275741">https://bugs.webkit.org/show_bug.cgi?id=275741</a>
<a href="https://rdar.apple.com/130288947">rdar://130288947</a>

Reviewed by Matt Woodrow.

The test would terminate GPU process.
Due WebRTC doing capture, it UI would restart the GPU process.
If WCP was in process of doing IPC sends, the termination would would
cause unresponsiveness event in WCP.
The unresponsiveness termination would intend to terminate the previous
process, but would actually terminate the new process.

Fix by associating the unresponsiveness call with
GPUProcessConnectionIdentifier.

* Source/WebKit/Scripts/webkit/messages.py:
(serialized_identifiers):
* Source/WebKit/Shared/WTFArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::createGPUProcessConnection):
(WebKit::WebProcessProxy::gpuProcessConnectionDidBecomeUnresponsive):
(WebKit::WebProcessProxy::gpuProcessExited):
* Source/WebKit/UIProcess/WebProcessProxy.h:
* Source/WebKit/UIProcess/WebProcessProxy.messages.in:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/GPU/GPUProcessConnection.h:
(WebKit::GPUProcessConnection::identifier const):
* Source/WebKit/WebProcess/WebProcess.cpp:
(WebKit::WebProcess::ensureGPUProcessConnection):
(WebKit::WebProcess::gpuProcessConnectionDidBecomeUnresponsive):

Canonical link: <a href="https://commits.webkit.org/280452@main">https://commits.webkit.org/280452@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/dff11fb883738aec1ba93882e70d10445ffb5717

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/56675 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/36002 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/9148 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/60289 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/7116 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/43625 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/7307 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/45909 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/4984 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/58705 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/33844 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/48915 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/26768 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/56201 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/30623 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/6246 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/6119 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/52566 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/6518 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/61972 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/587 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/6619 "Found 1 new test failure: webaudio/Panner/PannerNode-crash.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/53167 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/589 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/48973 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/53182 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/12537 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/505 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/31830 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/32916 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/34000 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/32662 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mattwoodrow commented on Source/WebKit/UIProcess/WebProcessProxy.cpp:
```diff
@@ -1080,11 +1080,14 @@ void WebProcessProxy::createGPUProcessConnection(IPC::Connection::Handle&& conne
     parameters.ignoreInvalidMessageForTesting = ignoreInvalidMessageForTesting();
 #endif
     parameters.isLockdownModeEnabled = lockdownMode() == WebProcessProxy::LockdownMode::Enabled;
+    m_gpuProcessConnectionIdentifier = identifier;
```
Can we assert that we don't already have an identifier?

It seems like sort of weird state to hold, since a WebProcess could in theory (but not in practice, currently) have multiple GPUProcessConnections.

Is there a guarantee that a GPUProcess crash would run `gpuProcessExited` in the UI process, before we receive a message from the WebProcess asking for a new connection (if that received the crash notification first)?

kkinnunen-apple replied:
> Can we assert that we don't already have an identifier?

Hmm, ok..

> It seems like sort of weird state to hold, since a WebProcess could in theory (but not in practice, currently) have multiple GPUProcessConnections.

We'd have list of known GPUProcessConnections. 

> Is there a guarantee that a GPUProcess crash would run `gpuProcessExited` in the UI process, before we receive a message from the WebProcess asking for a new connection (if that received the crash notification first)?

Yes, that's an existing race condition. The UI thinks the GPUP is still valid, and tries to send the new connections there.
However, nowadays the unresponsiveness notification should take care of it.

kkinnunen-apple committed:
REGRESSION (279833@main): webrtc/vp8-then-h264-gpu-process-crash.html is consistently crashing
https://bugs.webkit.org/show_bug.cgi?id=275741
rdar://130288947

Reviewed by Matt Woodrow.

The test would terminate GPU process.
Due WebRTC doing capture, it UI would restart the GPU process.
If WCP was in process of doing IPC sends, the termination would would
cause unresponsiveness event in WCP.
The unresponsiveness termination would intend to terminate the previous
process, but would actually terminate the new process.

Fix by associating the unresponsiveness call with
GPUProcessConnectionIdentifier.

* Source/WebKit/Scripts/webkit/messages.py:
(serialized_identifiers):
* Source/WebKit/Shared/WTFArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::createGPUProcessConnection):
(WebKit::WebProcessProxy::gpuProcessConnectionDidBecomeUnresponsive):
(WebKit::WebProcessProxy::gpuProcessExited):
* Source/WebKit/UIProcess/WebProcessProxy.h:
* Source/WebKit/UIProcess/WebProcessProxy.messages.in:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/GPU/GPUProcessConnection.h:
(WebKit::GPUProcessConnection::identifier const):
* Source/WebKit/WebProcess/WebProcess.cpp:
(WebKit::WebProcess::ensureGPUProcessConnection):
(WebKit::WebProcess::gpuProcessConnectionDidBecomeUnresponsive):

Canonical link: https://commits.webkit.org/280452@main
